src/middleware/persist.ts

```diff
@@ -0,0 +1,353 @@
+type EPersistentStorageName = string & { __isPersistentStorageName: true }
+type EPersistentStorageValue =
+  { state: ESelectedState
+  , version?: number & { __isPersistentStorageValueVersion: true } 
+  }
+
+type EPersist =
+    ( stateInitializer: StateInitializer<EState, Store<EState>>
+    , options: EPersistOptions
+    ) =>
+    StateInitializer<EState, Store<EState> & EPersistStore>
+
+interface EPersistOptions
+  { name: EPersistentStorageName
+  , getStorage?: () => EPersistentStorage
+  , serialize?:
+      (storageValue: EPersistentStorageValue) =>
+        MaybePromise<E.Serialized<EPersistentStorageValue>>
+  , deserialize?:
+      (serializedString: E.Serialized<EPersistentStorageValue>) =>
+        MaybePromise<EPersistentStorageValue>
+  , partialize?: (state: EState) => ESelectedState
+  , onRehydrateStorage?:
+      (state: EState) =>
+        void | ((state?: EState, error?: unknown) => void)
+  , version?: EPersistentStorageValue["version"]
+    migrate?:
+      ( persistedState: EPersistentStorageValue["state"]
+      , version: EPersistentStorageValue["version"]
+      ) =>
+        MaybePromise<EPersistentStorageValue["state"]>
+    merge?:
+      ( persistedState: EPersistentStorageValue["state"]
+      , currentState: EState
+      ) =>
+        EState
+  }
+
+interface EPersistentStorage
+  { getItem: (name: EPersistentStorageName) =>
+      MaybePromise<E.Serialized<EPersistentStorageValue> | null>
+  , setItem:
+      ( name: EPersistentStorageName
+      , value: E.Serialized<EPersistentStorageValue>
+      ) =>
+        MaybePromise<void>
+  , removeItem: (name: EPersistentStorageName) => MaybePromise<void>
+  }
+
+interface EPersistStore
+  { persist:
+      { setOptions: (options: O.Partial<EPersistOptions>) => void
+      , clearStorage: () => void
+      , rehydrate: () => Thenable<void>
+      , hasHydrated: () => boolean
+      , onHydrate: (listener: (state: EState) => void) => () => void
+      , onFinishHydration: (listener: (state: EState) => void) => () => void
+      }
+  }
+
+const persistImpl: EPersist = (stateInitializer, _options) => (parentSet, parentGet, parentStore) => {
+  let options = { ...defaultOptions, ..._options }
+
+  let persistentStorage = tryElse(options.getStorage, () => undefined)
+  let persistentStorageGetItem = () =>
+    persistentStorage!.getItem(options.name)
+  let persistentStorageSetItem = (serializedState: E.Serialized<EPersistentStorageValue>) =>
+    persistentStorage!.setItem(options.name, serializedState)
+  let persistentStorageRemoveItem = () =>
+    persistentStorage!.removeItem(options.name)
+  
+  if (!persistentStorage) {
+    return stateInitializer(
+      (...a) => {
+        console.warn(messages.unableToUpdateItem(options.name))
+        parentSet(...a);
+      },
+      parentGet, 
+      update(parentStore, "setState", setState => (...a) => {
+        console.warn(messages.unableToUpdateItem(options.name))
+        setState(...a);
+      })
```

Comment:
Previously this mutation was not made, I think it's good to be consistent and warn on both `setState`.

Also nowhere we're assuming that `set === store.set`, because they could be different functions (say while testing etc), so changes to `set` happen at two places.